### PR TITLE
Add a PORO for creating tax adjustments

### PR DESCRIPTION
**Pull request by mamhoff**
Adds a `Spree::Tax::ItemAdjuster`, which adjusts an item, and a `Spree::Tax::OrderAdjuster`, which iterates over line items and shipments in an order and adjusts those. 

This PR is a prerequisite for any further taxation work.

### CODE REVIEW

**Comment by mamhoff** [comment]
Force-pushed with no changes because of spurious failure in admin shipment spec.

**Comment by jhawthorn** [comment]
This does seem like a good target for extraction to a PORO, but the handling of the two cases (receiving either an item or an order) is a little awkward. The two cases share very little code.

Maybe we could adjust the signature of the adjuster? Brainstorming a little:

```
class Adjuster
  def initialize(order)
    @order = order
    @tax_zone = order.tax_zone
  end

  def adjust_order
    # adjust all items in the order
  end

  def adjust_item(item)
    # adjust just the item
  end
end
```

**Comment by tvdeyen** [comment]
One adjuster for two different things? It seams the only thing they share is the order tax zone. 

Why not an `OrderAdjuster` and a `LineItemAdjuster`, both knowing the order tax zone but having separated concerns?

Sandi Metz is calling! Excuse me ;)

**Comment by tvdeyen** [comment]
There is too much knowledge of the two possible callers inside one class. Case statements on class types, then methods named after the caller, etc. 

Two separate adjusters are much more flexible, more future safe and less error prone. If we use POROs than please let us use clean object oriented design. 

Sorry, reading too much Sandi Metz lately. 

Just my 0.02

**Comment by mamhoff** [comment]
Thank you for chipping in! 

@jhawthorn Actually, they do share code, as the order adjuster doesn't actually adjust the order, but all the items (line items and shipments) contained within it. 

The current stuff on `TaxRate` shaves some AR requests by only loading the `potentially_applicable` rates once per order - I would like that efficiency to stay, which is why I didn't separate the two into their own classes. @tvdeyen I hear you though. Possibly, though, I can have the best of both worlds by making the `LineItemAdjuster` a Singleton and cache things like the tax rates in an instance variable. 

At the moment, `Spree::TaxRate.adjust` is a method that is executed for side effects (create adjustments, update `pre_tax_amount`), which isn't super transparent. Ideally, I'd like something to this effect: 

```
class Spree::LineItem
  has_many :adjustments, ... , autosave: true
  before_save :adjust

  private 
  
  # builds tax adjustments and sets pre tax amount, but doesn't actually save
  def adjust
    self = Spree::Tax::ItemAdjuster.new(self).adjust!
  end
end
```

**Comment by mamhoff** [comment]
Changed these to single responsibility objects - I do think we can deal with the efficiency after we've defined the interfaces.

File: core/app/models/spree/order.rb [hunk 1]
```diff
@@ -324,10 +324,7 @@ def line_item_options_match(line_item, options)
     # Creates new tax charges if there are any applicable rates. If prices already
     # include taxes then price adjustments are created instead.
     def create_tax_charge!
-      # We want to only look up the applicable tax zone once and pass it to TaxRate calculation to avoid duplicated lookups.
-      order_tax_zone = self.tax_zone
-      Spree::TaxRate.adjust(order_tax_zone, line_items)
-      Spree::TaxRate.adjust(order_tax_zone, shipments) if shipments.any?
+      Spree::Tax::Adjuster.new(self).adjust!
```
Comment (tvdeyen) [inline core/app/models/spree/order.rb]: Don't you want to call `OrderAdjuster` here?

Comment (mamhoff) [inline core/app/models/spree/order.rb]: Absolutely. Thanks for the heads-up! :100:

**Comment by mamhoff** [comment]
Wheew! Green! Removing `TaxRate.adjust`, which is a monster, has not been that hard after all. I'm excited: I like the API, and because of some data sharing between the `ItemAdjuster` and the `OrderAdjuster` the expensive `TaxRate.match` call is only done once. 

This is still lacking feedback. Once that's in, I'd start putting in unit tests and moving the Spree::TaxRate into some kind of `taxation_integration_spec.rb` file. 

Please please, this isn't ready for merge, but definitely ready for feedback (although you do have a couple of days as I'll be out until the weekend from tomorrow).

**Comment by mamhoff** [comment]
Rebased on master, updated commit message on second commit. Feedback still welcome. 

Btw: With #530 in, I'll take a go at removing `TaxRate.match`, too.

**Comment by mamhoff** [comment]
These last ones failed because of the Rails update (https://github.com/solidusio/solidus/pull/736). Rebasing on master now, and adding a couple of performance fixes as well.

**Comment by mamhoff** [comment]
The build failed because of phantomjs crashing :(

**Comment by mamhoff** [comment]
Force-pushed to get a green build. Otherwise: This is ready for review. 

I can think of more stuff to do, but the scope of this PR is already quite large, so I'll stop for now. :)

**Comment by peterberkenbosch** [comment]
:+1: Nice @mamhoff

**Comment by mamhoff** [comment]
Damn. The shipping rates also have VAT, and they're not price-adjusted yet.

**Comment by tvdeyen** [comment]
I hopefully can spent some review time later this evening (CET)

**Comment by mamhoff** [comment]
This is really long, and hard to review. I'll split this PR in two: One that just adds the POROs and uses #530 to look up zones (that's all commits up to a68d9c8), and one that actually implements MOSS with that. 

The one that actually implements MOSS will also have to deal with shipping rates, somehow.

**Comment by tvdeyen** [comment]
Alright

File: core/app/models/spree/tax/item_adjuster.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
+module Spree
+  module Tax
+    class ItemAdjuster
+      include TaxHelpers
+
+      def initialize(item, options = {})
+        @item = item
+        @order = @item.order
```
Comment (tvdeyen) [inline core/app/models/spree/tax/item_adjuster.rb]: `@item` and `@order` should be `attr_reader`s, so the `TaxHelper` `module` and potential sub classes don't rely on reading instance variables.

File: core/app/models/spree/tax/item_adjuster.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        # set the instance variable so `TaxRate.match` is only called when necessary
+        @order_rates = options[:order_tax_rates]
+      end
+
+      def adjust!
+        return unless order_tax_zone
+        # Using .destroy_all to make sure callbacks fire
+        @item.adjustments.tax.destroy_all
```
Comment (tvdeyen) [inline core/app/models/spree/tax/item_adjuster.rb]: after introduced we should use the `item` `attr_reader` here and not use instance variables.

File: core/app/models/spree/tax/item_adjuster.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+        rates_for_item.each do |rate|
+          rate.adjust(order_tax_zone, @item)
+        end
+      end
+
+      private
+
+      def rates_for_item
+        @rates ||= applicable_rates.select { |rate| rate.tax_category == @item.tax_category }
```
Comment (tvdeyen) [inline core/app/models/spree/tax/item_adjuster.rb]: I find it good practice to name caching instance variables after the methods name and use an underscore to make it more visible, that this is not a normal instance variable. So, the name should be `@_rates_for_item` IMO. 

And we should also use the `item` `attr_reader` here.

File: core/app/models/spree/tax/order_adjuster.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+module Spree
+  module Tax
+    class OrderAdjuster
+      include TaxHelpers
+
+      def initialize(order)
+        @order = order
```
Comment (tvdeyen) [inline core/app/models/spree/tax/order_adjuster.rb]: Same as in item adjuster. Use a `attr_reader` for `order`

File: core/app/models/spree/tax/tax_helpers.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      def rates_for_order_zone
+        @rates_for_order_zone ||= Spree::TaxRate.for_zone(order_tax_zone)
+      end
+
+      def rates_for_default_zone
+        @rates_for_default_zone ||= Spree::TaxRate.for_zone(Spree::Zone.default_tax)
+      end
+
+      def order_tax_zone
+        @order_tax_zone ||= @order.tax_zone
```
Comment (tvdeyen) [inline core/app/models/spree/tax/tax_helpers.rb]: This is exactly why we should use the `order` `attr_reader`. The `module` shouldn't know anything about an instance variable set in the including class.

File: core/app/models/spree/tax/tax_helpers.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+        (rates_for_order_zone + default_rates_with_unmatched_tax_category).uniq
+      end
+
+      def rates_for_order_zone
+        @rates_for_order_zone ||= Spree::TaxRate.for_zone(order_tax_zone)
+      end
+
+      def rates_for_default_zone
+        @rates_for_default_zone ||= Spree::TaxRate.for_zone(Spree::Zone.default_tax)
```
Comment (tvdeyen) [inline core/app/models/spree/tax/tax_helpers.rb]: Here the same rule for naming caching instance variables applies

File: core/app/models/spree/tax/tax_helpers.rb (hunk 6 above)
Comment (tvdeyen) [inline core/app/models/spree/tax/tax_helpers.rb]: also caching variable name

File: core/spec/models/spree/tax/item_adjuster_spec.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  let(:item) { Spree::LineItem.new(order: order) }
+
+  before do
+    allow(order).to receive(:tax_zone) { build(:zone) }
+  end
+
+  describe 'initialization' do
+
+    it 'sets @order to adjustable.order' do
+      expect(adjuster.instance_variable_get(:@order)).to eq(item.order)
```
Comment (tvdeyen) [inline core/spec/models/spree/tax/item_adjuster_spec.rb]: A prime example why we need the `order` `attr_reader`

Comment (tvdeyen) [inline core/spec/models/spree/tax/item_adjuster_spec.rb]:
```ruby
expect(adjuster.order).to eq(item.order)
```

:heart_eyes:

**Comment by mamhoff** [comment]
Did the naming changes, and also added some more tests to the adjusters.

**Comment by mamhoff** [comment]
First commit was failing because of a naming change, rebased again.

File: core/app/models/spree/tax/item_adjuster.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    class ItemAdjuster
+      attr_reader :item, :order
+
+      include TaxHelpers
+
+      def initialize(item, options = {})
+        @item = item
+        @order = @item.order
+        # set the instance variable so `TaxRate.match` is only called when necessary
+        @rates_for_order_zone = options[:rates_for_order_zone]
```
Comment (gmacdougall) [inline core/app/models/spree/tax/item_adjuster.rb]: I'm not clear where this is being used.

File: core/app/models/spree/tax/item_adjuster.rb [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+      end
+
+      def adjust!
```
Comment (gmacdougall) [inline core/app/models/spree/tax/item_adjuster.rb]: Could we add some YARD here to explain the return values of this method.

I also tend to to not have a method ending in a bang without a corresponding non-bang method.  See http://dablog.rubypal.com/2007/8/15/bang-methods-or-danger-will-rubyist

File: core/app/models/spree/tax/order_adjuster.rb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+      end
+
+      def adjust!
+        return unless order_tax_zone
+
+        (order.line_items + order.shipments).each do |item|
```
Comment (gmacdougall) [inline core/app/models/spree/tax/order_adjuster.rb]: Is this a concept we should have on the order model? Some sort of method that returns all things that can be adjusted?

File: core/app/models/spree/zone.rb [hunk 12]
```diff
@@ -58,6 +58,8 @@ def self.match(address)
     # current zone, if it's a state zone. If the passed-in zone has members, it
     # will also be in the result set.
     def self.with_shared_members(zone)
+      return none unless zone
```
Comment (gmacdougall) [inline core/app/models/spree/zone.rb]: Does this work?

Comment (jhawthorn) [inline core/app/models/spree/zone.rb]: Looks like it should to me. This is a class method so it will return the `Spree::Zone.none` null object relation.

Comment (mamhoff) [inline core/app/models/spree/zone.rb]: What John said. However, medium term that entire method goes away when #783 gets applied.

File: core/app/models/spree/tax/order_adjuster.rb (hunk 11 above)
Comment (mamhoff) [inline core/app/models/spree/tax/order_adjuster.rb]: We could, potentially. I shied away of adding more methods to `Spree::Order`. I'd do that on a later commit though once we actually see it DRYing away other things, namely when I have to write a method that does the same thing for the shipping rates in a shipment (https://github.com/solidusio/solidus/pull/759).

File: core/app/models/spree/tax/item_adjuster.rb (hunk 10 above)
Comment (mamhoff) [inline core/app/models/spree/tax/item_adjuster.rb]: I can very well imagine a future non-bang `adjust` method: One that will not change the order, and not create adjustments, but just `build` them. That would be great to reduce database calls when updating items and orders. Alas, we're not there yet, but I'd like to keep this method's name the same at that point. 

YARD I can absolutely add, as this method returns the newly created adjustments (making it somewhat easier to test than `Spree::TaxRate.adjust`).

File: core/app/models/spree/tax/item_adjuster.rb (hunk 9 above)
Comment (mamhoff) [inline core/app/models/spree/tax/item_adjuster.rb]: These are used in private methods of the included module `TaxHelpers`, in order to run the database calls that only have to run once per order only once per order: https://github.com/magiclabs/solidus/blob/add-tax-adjuster/core/app/models/spree/tax/tax_helpers.rb#L28-L38.

**Comment by jhawthorn** [comment]
Looks good to me. :+1:

One thought, should we keep `TaxRate.adjust` as a deprecated method that would use the PORO?

**Comment by mamhoff** [comment]
Well, the method signatures are a bit different. I could use just the item adjuster, but then that method will have abysmal performance. I'd rather have it entirely gone. However, if there's another person saying: Deprecate it and be nice about it - I'll do it, along with the requested YARD docs tomorrow.

File: core/app/models/spree/zone.rb (hunk 12 above)
Comment (gmacdougall) [inline core/app/models/spree/zone.rb]: I had forgotten that `none` is an ActiveRecord method. Looked for our own implementation and didn't see one, and got confused. Sorry spreading the confusion.

**Comment by tvdeyen** [comment]
:icecream:

**Comment by timmyArch** [comment]
Nice ;)

**Pull request merged**